qiskit/pulse/builder.py

```diff
@@ -0,0 +1,1996 @@
+        """
+        #: BaseBackend: Backend instance for context builder.
+        self._backend = backend
+
+        #: Union[None, ContextVar]: Token for this ``_PulseBuilder``'s ``ContextVar``.
+        self._backend_ctx_token = None
+
+        #: pulse.Schedule: Active schedule of BuilderContext.
+        self._block = None
+
+        #: QuantumCircuit: Lazily constructed quantum circuit
+        self._lazy_circuit = None
+
+        # ContextManager: Default alignment context.
+        self._default_alignment_context = _align(default_alignment)
+
+        if default_transpiler_settings is None:
+            default_transpiler_settings = {}
+        self._transpiler_settings = default_transpiler_settings
+
+        if default_circuit_scheduler_settings is None:
+            default_circuit_scheduler_settings = {}
+        self._circuit_scheduler_settings = default_circuit_scheduler_settings
+
+        if schedule is None:
+            schedule = Schedule()
+        # pulse.Schedule: Root program block
+        self._schedule = schedule
+
+        self.set_active_block(Schedule())
+
+    def __enter__(self) -> Schedule:
+        """Enter this builder context and yield either the supplied schedule
+        or the schedule created for the user.
+
+        Returns:
+            The schedule that the builder will build on.
+        """
+        self._backend_ctx_token = BUILDER_CONTEXTVAR.set(self)
+        self._default_alignment_context.__enter__()
+        return self._schedule
+
+    @_compile_lazy_circuit_before
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """Exit the builder context and compile the built pulse program."""
+        self._default_alignment_context.__exit__(exc_type, exc_val, exc_tb)
+        self.compile()
+        BUILDER_CONTEXTVAR.reset(self._backend_ctx_token)
+
+    @property
+    def backend(self):
+        """Returns the builder backend if set.
+
+        Returns:
+            Optional[BaseBackend]: The builder's backend.
+        """
+        return self._backend
+
+    @property
+    def block(self) -> Schedule:
+        """Return the active block of this bulder."""
+        return self._block
+
+    @property
+    @_requires_backend
+    def num_qubits(self):
+        """Get the number of qubits in the backend."""
+        return self.backend.configuration().n_qubits
+
+    @property
+    def transpiler_settings(self) -> Mapping:
+        """The builder's transpiler settings."""
+        return self._transpiler_settings
+
+    @transpiler_settings.setter
+    @_compile_lazy_circuit_before
+    def transpiler_settings(self, settings: Mapping):
+        self._compile_lazy_circuit()
+        self._transpiler_settings = settings
+
+    @property
+    def circuit_scheduler_settings(self) -> Mapping:
+        """The builder's circuit to pulse scheduler settings."""
+        return self._circuit_scheduler_settings
+
+    @circuit_scheduler_settings.setter
+    @_compile_lazy_circuit_before
```

Comment:
If the settings change it will override the previous settings that the circuit was added under. for example if the user did 

```python
with pulse.circuit_scheduler_settings(setting=1):
     pulse.call(circuit1)
     with pulse.circuit_scheduler_settings(setting=2):
            pulse.call(circuit2)
```
I would assume the first settings would be used for circuit 1 and the second for setting 2. If we don't force the compile between settings this won't be the case.